clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy

```diff
@@ -80,50 +80,77 @@ abstract class OpenstackClientProvider {
-      def securityGroupsApi = client.compute().securityGroups()
+  /**
+   * Deletes a security group rule
+   * @param region the region to delete the rule from
+   * @param id id of the rule to delete
+   */
+  void deleteSecurityGroupRule(String region, String id) {
+    handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
+      client.useRegion(region).compute().securityGroups().deleteRule(id)
+    }
+  }
 
-      // Try getting existing security group, update if needed
-      SecGroupExtension securityGroup
-      if (StringUtils.isNotEmpty(securityGroupId)) {
-        securityGroup = securityGroupsApi.get(securityGroupId)
-      }
-      if (securityGroup == null) {
-        securityGroup = securityGroupsApi.create(securityGroupName, description)
-      } else {
-        securityGroup = securityGroupsApi.update(securityGroup.id, securityGroupName, description)
-      }
+  /**
+   * Updates a security group with the new name and description
+   * @param region the region the security group is in
+   * @param id the id of the security group to update
+   * @param name the new name for the security group
+   * @param description the new description for the security group
+   * @return the updated security group
+   */
+  SecGroupExtension updateSecurityGroup(String region, String id, String name, String description) {
+    handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
+      client.useRegion(region).compute().securityGroups().update(id, name, description)
+    }
+  }
 
-      // TODO: Find the different between existing rules and only apply that instead of deleting and re-creating all the rules
-      securityGroup.rules.each { rule ->
-        securityGroupsApi.deleteRule(rule.id)
-      }
+  /**
+   * Creates a security group with the given name and description
+   * @return the created security group
+     */
+  SecGroupExtension createSecurityGroup(String region, String name, String description) {
+    handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
+      client.useRegion(region).compute().securityGroups().create(name, description)
+    }
+  }
 
-      rules.each { rule ->
-        securityGroupsApi.createRule(Builders.secGroupRule()
-          .parentGroupId(securityGroup.id)
-          .protocol(IPProtocol.valueOf(rule.ruleType))
-          .cidr(rule.cidr)
-          .range(rule.fromPort, rule.toPort).build())
+  /**
+   * Returns the security group for the given id.
+   * @param region the region to look up the security group in
+   * @param id id of the security group.
+     */
+  SecGroupExtension getSecurityGroup(String region, String id) {
+    handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
+      SecGroupExtension securityGroup = client.useRegion(region).compute().securityGroups().get(id)
+      if (securityGroup == null) {
+        throw new OpenstackProviderException("Unable to find security group ${id}")
```

Comment:
Spoke with Ashley on Slack about this. We started splitting up the exceptions to `OpenstackProviderExeception`s and `OpenstackOperationException`s as it seems wrong to throw an operation exception when trying to get a list of instances for caching. That new exception subclasses `OpenstackOperationException`.

To allow us to have a worthwhile `OpenstackResourceNotFoundException`, I am going to ahve it subclass the provider exception and check for null there instead of the operation.